src/librustc_mir/interpret/validity.rs

```diff
@@ -201,13 +201,16 @@ impl<'rt, 'mir, 'tcx, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, 'tcx, M
 
             // enums
             ty::Adt(def, ..) if def.is_enum() => {
-                // we might be projecting *to* a variant, or to a field *in*a variant.
+                // we might be projecting *to* a variant, or to a field *in* a variant.
                 match layout.variants {
                     layout::Variants::Single { index } => {
                         // Inside a variant
                         PathElem::Field(def.variants[index].fields[field].ident.name)
                     }
-                    _ => bug!(),
+                    layout::Variants::Multiple { discr_index, .. } => {
+                        assert_eq!(discr_index, field);
+                        PathElem::Tag
+                    }
```

Comment:
Handling this case (specifically `field == discr_index`) can probably be made more general, and checked first, so it works with generators too (right now generators get `PathElem::ClosureVar(sym::integer(discr_index))` in that case instead of `PathElem::Tag`).

Reply:
All right, I will add such a general case above.